tests/introspection/ApiPInvokeTest.cs

```diff
@@ -238,15 +238,109 @@ protected void Check (Assembly a)
 
 					var lib = Dlfcn.dlopen (path, 0);
 					var h = Dlfcn.dlsym (lib, name);
-					if (h == IntPtr.Zero)
+					if (h == IntPtr.Zero) {
 						ReportError ("Could not find the symbol '{0}' in {1} for the P/Invoke {2}.{3} in {4}", name, path, t.FullName, m.Name, a.GetName ().Name);
+					} else if (path != null) {
+						// Verify that the P/Invoke points to the right library.
+						Dl_info info = default (Dl_info);
+						var found = dladdr (h, out info);
+						if (found != 0) {
+							// Resolve symlinks in both cases
+							var dllImportPath = ResolveLibrarySymlinks (path);
+							var foundLibrary = ResolveLibrarySymlinks (Marshal.PtrToStringAuto (info.dli_fname));
+							if (Skip (name, ref dllImportPath, ref foundLibrary)) {
+								// Skipped
+							} else if (foundLibrary != dllImportPath) {
```

Comment:
Why not merging the if statement:
```suggestion
							if (!Skip (name, ref dllImportPath, ref foundLibrary) && (foundLibrary != dllImportPath) {
```

Reply:
Easier to add CWLs for debugging (I also find it easier to read).